src/main/java/com/google/sps/servlets/KeywordServlet.java

```diff
@@ -17,12 +25,36 @@
 @WebServlet("/keyword")
 public class KeywordServlet extends HttpServlet {
 
-  private static final List<String> tweets = new ArrayList<>();
-  private static final List<String> terms = Arrays.asList("Black Lives Matter", "COVID-19");
+  private final List<String> tweets = new ArrayList<>();
+  private final TreeMap<Entity, Float> terms = new TreeMap<>((term1, term2) -> {
+    if (term1.getSalience() > term2.getSalience()) {
+      return 1;
+    } else if (term1.getSalience() < term2.getSalience()) {
+      return -1;
+    } else {
+      return 0;
+    }
+  });
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // TODO: Extract saliency.
+    for (String tweet : tweets) {
+      AnalyzeEntitiesResponse entityResponse = analyzeEntity(tweet);
+      List<Entity> entities = entityResponse.getEntitiesList();
+      for (Entity entity : entities) {
+        if (entity.getSalience() == 0) {
+          continue;
+        }
+        
+        if (terms.size() < 10) {
+          terms.put(entity, entity.getSalience());
+        } else if (!terms.containsKey(entity) && entity.getSalience() > terms.get(terms.pollFirstEntry())) {
```

Comment:
Why do you need to check here if terms doesn't contain the entity but not in the if statement on line 49. If analyzeEntity only generates unique entities, you shouldn't have to check `!terms.containsKey(entity)` here. Otherwise, you would need to check `!terms.containsKey(entity)` before the previous if statement as well.

Reply:
This is mainly for considering multiple tweets. If the user has, say, 2 tweets that mentions "Black Lives Matter" for example, then there would be two entities with the same word. However, I'm not exactly sure how the entity hashes, as it could be the case that two different entities with the same _name_ field could hash to two completely different things.

Reply:
Ah, I guess I didn't really read the question. In the first _if_ statement, it wouldn't matter as there are less than 10 terms (for now, we're only looking at the top 10 keywords). In addition, if the hashCode() method only considers equality in _name_, then this would also simply just replace the previous salience score with the new one. However, this may or may not be desired at all, so I will add something to account for that.

Reply:
Ok, that makes sense to me.